projects/email_alert_notifications/resources/email_alert_notifications.tf

```diff
@@ -0,0 +1,63 @@
+variable "s3_bucket_name" {
+  default = "govuk-email-alert-notifications"
+}
+
+resource "template_file" "s3_bucket_policy" {
+  template = "${file("templates/email_alert_s3_bucket_policy.json")}"
+  vars {
+    account_id = "${element(split(":", aws_iam_role.lambda_execute_and_write_to_email_alert_bucket.arn), 4)}"
+    bucket_name = "${var.s3_bucket_name}"
+    lambda_role = "${aws_iam_role.lambda_execute_and_write_to_email_alert_bucket.arn}"
+  }
+}
+
+resource "template_file" "put_and_delete_to_email_alert_bucket_policy" {
+  template = "${file("templates/put_and_delete_to_s3_policy.json")}"
+  vars {
+    resource_arn = "${aws_s3_bucket.email_alert_inbox_bucket.arn}"
+  }
+}
+
+resource "aws_s3_bucket" "email_alert_inbox_bucket" {
+  bucket = "${var.s3_bucket_name}"
+  acl = "public-read"
+  policy = "${template_file.s3_bucket_policy.rendered}"
+}
+
+resource "aws_iam_role" "lambda_execute_and_write_to_email_alert_bucket" {
+  name = "lambda_execute_and_write_to_email_alert_bucket" 
+  assume_role_policy = "${file("templates/lambda_assume_role_policy.json")}"
+}
+
+resource "aws_iam_role_policy" "put_and_delete_to_email_alert_bucket" {
+  name = "put_and_delete_to_email_alert_bucket"
+  role = "${aws_iam_role.lambda_execute_and_write_to_email_alert_bucket.id}"
+  policy = "${template_file.put_and_delete_to_email_alert_bucket_policy.rendered}"
+}
+
+resource "aws_iam_role_policy" "write_to_logs" {
+  name = "write_to_logs"
+  role = "${aws_iam_role.lambda_execute_and_write_to_email_alert_bucket.id}"
+  policy = "${file("templates/write_to_logs_policy.json")}"
+}
+
+resource "aws_lambda_function" "rename_email_files_with_request_id" {
+  depends_on = ["null_resource.build_rename_email_files_with_request_id"]
+  filename = "files/rename_email_files_with_request_id.zip"
+  function_name = "rename_email_files_with_request_id"
+  role = "${aws_iam_role.lambda_execute_and_write_to_email_alert_bucket.arn}"
+  handler = "rename_email_files_with_request_id.lambda_handler"
+  runtime = "python2.7"
+  source_code_hash = "${base64sha256(file("files/rename_email_files_with_request_id.py"))}"
```

Comment:
> Is there a better way that I can trigger this resource to be updated when the .py source changes?

Before Terraform gets to `plan/apply` stage, it (by default) performs `refresh` (effectively `Read` from CRUD for each resource in all `*.tf` files) so it can figure out what may have changed and how does it affect `plan/apply`. After refresh is done for all resources, it performs `plan/apply`.

> As a result the base64sha256 function raises a no such file error if I use the zip file.

The `base64sha256()` function is executed at the time of `refresh` since it has static value (static path to zip file) hence Terraform won't mark it as "computed". All provisioners are always executed at the time of `apply`, not during `refresh`. That's why you're hitting this.

To mark it as "computed" you would probably need to have a 1st class support for "zipping" in terraform which would be returning the shasum that you would reference in `aws_lambda_function. source_code_hash`. There are some (good) reasons why Terraform doesn't support this (yet):
https://github.com/hashicorp/terraform/pull/3858

The only way to get around this would be possibly executing terraform twice (+ removing any unnecessary `triggers` & `depends_on`):

```
terraform apply -target=null_resource.build_rename_email_files_with_request_id
terraform plan/apply
```

that way terraform won't perform any validation/checks on any resources except the targeted one.
Obviously the `null_resource`'s provisioner would be executed twice (in the second run too), but at least Terraform would get a chance to prepare the right `plan` and `apply` it correctly.

Also keep in mind that any `local-exec` provisioner makes your Terraform files platform-specific - i.e. the requirement suddenly isn't just terraform binary, but also platform-specific `zip` utility having the same interface (e.g. supporting `-j` flag). This is also why I think that building ZIP files via CI is a better solution.